src/ess/reduce/nexus.py

```diff
@@ -0,0 +1,329 @@
+    group_name: Optional[str],
+    nx_class: Type[snx.NXobject],
+    instrument_name: Optional[InstrumentName] = None,
+) -> sc.DataGroup:
+    with _open_nexus_file(file_path) as f:
+        entry = f['entry']
+        instrument = _unique_child_group(entry, snx.NXinstrument, instrument_name)
+        loaded = cast(
+            sc.DataGroup, _unique_child_group(instrument, nx_class, group_name)[()]
+        )
+        loaded = snx.compute_positions(loaded)
+        return loaded
+
+
+def _open_nexus_file(
+    file_path: Union[FilePath, NeXusFile, NeXusGroup]
+) -> ContextManager:
+    if isinstance(file_path, getattr(NeXusGroup, '__supertype__', type(None))):
+        return nullcontext(file_path)
+    return snx.File(file_path)
+
+
+def _unique_child_group(
+    group: snx.Group, nx_class: Type[snx.NXobject], name: Optional[str]
+) -> snx.Group:
+    if name is not None:
+        child = group[name]
+        if isinstance(child, snx.Field):
+            raise ValueError(
+                f"Expected a NeXus group as item '{name}' but got a field."
+            )
+        if child.nx_class != nx_class:
+            raise ValueError(
+                f"The NeXus group '{name}' was expected to be a "
+                f'{nx_class} but is a {child.nx_class}.'
+            )
+        return child
+
+    children = group[nx_class]
+    if len(children) != 1:
+        raise ValueError(f'Expected exactly one {nx_class} group, got {len(children)}')
+    return next(iter(children.values()))  # type: ignore[return-value]
+
+
+def extract_detector_data(
+    detector: RawDetector, detector_name: DetectorName
+) -> RawDetectorData:
+    """Get and return the events or histogram from a detector loaded from NeXus.
+
+    Parameters
+    ----------
+    detector:
+        A detector loaded from NeXus.
+    detector_name:
+        Name of the detector.
+
+    Returns
+    -------
+    :
+        A data array containing the events or histogram.
+
+    See also
+    --------
+    load_detector:
+        Load a detector from a NeXus file in a format compatible with
+        ``extract_detector_data``.
+    """
+    return RawDetectorData(_extract_events_or_histogram(detector, detector_name))
+
+
+def extract_monitor_data(
+    monitor: RawMonitor, monitor_name: MonitorName
+) -> RawMonitorData:
+    """Get and return the events or histogram from a monitor loaded from NeXus.
+
+    Parameters
+    ----------
+    monitor:
+        A monitor loaded from NeXus.
+    monitor_name:
+        Name of the monitor.
+
+    Returns
+    -------
+    :
+        A data array containing the events or histogram.
+
+    See also
+    --------
+    load_monitor:
+        Load a monitor from a NeXus file in a format compatible with
+        ``extract_monitor_data``.
+    """
+    return RawMonitorData(_extract_events_or_histogram(monitor, monitor_name))
+
+
+def _extract_events_or_histogram(dg: sc.DataGroup, name: str) -> sc.DataArray:
+    data_names = {f'{name}_events', 'data'}
```

Comment:
I suggest to remove the reliance on naming. Just look for child data arrays, prefer to return the event-data one?